framework/src/org/checkerframework/common/jaifinference/JaifInferenceUtils.java

```diff
@@ -0,0 +1,485 @@
+                return setOfAnnotationsToATM(annos, atf, false, fieldNode.getType());
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        return null;
+    }
+
+    /**
+     * Updates the type of the field lhs in the .jaif file of the class with
+     * symbol classSymbol. The new type will be the LUB between the previous
+     * type and the type of lhs. If the new type is not a subtype of the type of
+     * lhs, no refinement is made.
+     * @param lhs is the node representing the field.
+     * @param rhs is the node representing the RHS of the assignment.
+     * @param classSymbol is the symbol of the class representing the class 
+     * that contains the field.
+     * @param atf is the annotated type factory.
+     */
+    public static void updateFieldTypeInJaif(FieldAccessNode lhs, Node rhs,
+            ClassSymbol classSymbol, AnnotatedTypeFactory atf) {
+        if (classSymbol == null) return;
+        String jaifPath = JAIF_FILES_PATH + classSymbol.flatname.toString().
+                replaceAll("\\.", "/") + ".jaif";
+        AnnotatedTypeMirror rhsATM = atf.getAnnotatedType(rhs.getTree());
+        AnnotatedTypeMirror lhsATM = atf.getAnnotatedType(lhs.getTree());
+        if (lhsATM.getExplicitAnnotations().size() > 0) {
+            // We do not infer types if there are explicit annotations.
+            // See https://github.com/typetools/annotation-tools/issues/105
+            return;
+        }
+        AnnotatedTypeMirror prevATM = null;
+
+        AScene scene = new AScene();
+        try {
+            AClass clazz = getJaifClass(classSymbol, jaifPath, scene);
+            if (clazz == null) return;
+            String fieldName = lhs.getFieldName();
+            AField field = clazz.fields.get(fieldName);
+            if (field == null) {
+                field = clazz.fields.vivify(fieldName);
+            }
+            Set<Annotation> prevAnnos = field.tlAnnotationsHere;
+            if (prevAnnos != null && prevAnnos.size() > 0) {
+                prevATM = setOfAnnotationsToATM(prevAnnos, atf, false,
+                        rhsATM.getUnderlyingType());
+                if (prevATM != null) {
+                    // If there was a type previously, we must do the LUB to keep soundness.
+                    rhsATM = AnnotatedTypes.leastUpperBound(atf.getProcessingEnv(),
+                            atf, rhsATM, prevATM);
+                }
+            }
+            // Write into .jaif file ONLY IF refined type is a subtype of the current type.
+            if (atf.getTypeHierarchy().isSubtype(rhsATM, lhsATM)) {
+                Set<Annotation> setOfAnnos = atmToSetOfAnnotations(rhsATM, atf);
+                field.tlAnnotationsHere.clear();
+                field.tlAnnotationsHere.addAll(setOfAnnos);
+                IndexFileWriter.write(scene, new FileWriter(jaifPath));
+            }
+        } catch (IOException e) {
+            e.printStackTrace();
```

Comment:
Similar questions to above about suppressing fatal errors, which is poor practice.